packages/next/src/server/stream-utils/node-web-streams-helper.ts

```diff
@@ -273,45 +283,41 @@ function createMergedTransformStream(
     // implementation, e.g. with a specific high-water mark. To ensure it's
     // the safe timing to pipe the data stream, this extra tick is
     // necessary.
-    const detached = new DetachedPromise<void>()
-    pending = detached
 
-    // We use `setTimeout/setImmediate` here to ensure that it's inserted after
-    // flushing the shell. Note that this implementation might get stale if impl
-    // details of Fizz change in the future.
-    scheduleImmediate(async () => {
-      try {
-        while (true) {
-          const { done, value } = await reader.read()
-          if (done) return
+    // We don't start reading until we've left the current Task to ensure
+    // that it's inserted after flushing the shell. Note that this implementation
+    // might get stale if impl details of Fizz change in the future.
+    await atLeastOneTask()
 
-          controller.enqueue(value)
+    try {
+      while (true) {
+        const { done, value } = await reader.read()
+        if (done) {
+          donePulling = true
+          return
         }
-      } catch (err) {
-        controller.error(err)
-      } finally {
-        detached.resolve()
+
+        controller.enqueue(value)
       }
-    })
+    } catch (err) {
+      controller.error(err)
+    }
   }
 
   return new TransformStream({
     transform(chunk, controller) {
       controller.enqueue(chunk)
 
       // Start the streaming if it hasn't already been started yet.
-      if (started) return
-      started = true
-
-      start(controller)
+      if (!pull) {
+        pull = startPulling(controller)
+      }
```

Comment:
NIT: Early returns here might make this easier to read:

```suggestion
      if (pull) return
      pull = startPulling(controller)
```

Reply:
If this code grows to such a complexity that an early return offers clarity by scoping the environment to some invariant I'm all for it but it's so trivial I don't think it reads any clearer